### PR TITLE
Refactor: separate HTTP handlers from app logic (galley)

### PR DESCRIPTION
**Pull request by mheinzel**
This has already been done for brig in #967, now for galley.

### CODE REVIEW

**Comment by mheinzel** [comment]
There are integration test failures because of status code mismatches, I'll update this PR once it's fixed.

**Comment by mheinzel** [comment]
Passing now.

I don't like the fact that we have endpoints that only in some cases require a JSON body (as well as potentially masking invalid-payload errors by permission errors), but I can't change that. At least using `OptionalJsonRequest` seems to be the proper way of specifying this behaviour.

File: services/galley/src/Galley/API/LegalHold.hs [hunk 1]
```diff
@@ -131,8 +156,18 @@ getUserStatus (_zusr ::: tid ::: uid ::: _) = do
...
 -- | Request to provision a device on the legal hold service for a user
-requestDevice :: UserId ::: TeamId ::: UserId ::: JSON -> Galley Response
-requestDevice (zusr ::: tid ::: uid ::: _) = do
+requestDeviceH :: UserId ::: TeamId ::: UserId ::: JSON -> Galley Response
+requestDeviceH (zusr ::: tid ::: uid ::: _) = do
+  requestDevice zusr tid uid <&> \case
+    RequestDeviceSuccess -> empty & setStatus status201
+    RequestDeviceAlreadyPending -> empty & setStatus status204
+
+data RequestDeviceResult
```
Comment (arianvp) [inline services/galley/src/Galley/API/LegalHold.hs]: If it's a response I think it should be in `galley-types` no?

**Comment by mheinzel** [comment]
> At least using OptionalJsonRequest seems to be the proper way of specifying this behaviour.

To clarify: The current approach DOES slightly change behaviour if the request body both
- contains something, but not a valid `TeamDeleteData`/`TeamMemberDeleteData` JSON structure
- is not looked at by the endpoint in that specific branch of execution (i.e. non-binding or non-existing team or not having necessary permission)

For non-binding teams, we should check what clients send. Do they ever make a request knowing that this is a non-binding team, so they leave out the body or maybe send an empty object? (especially with non-binding teams being a legacy thing)

The other cases seems like a client bug to me. I would rather find out that there is invalid JSON sent than getting an permission error. Also, if this happened, it would most likely also happen in the case of having proper permission, so I don't think this is hiding somewhere in the the current clients.

That being said, maybe we don't want to change behaviour to avoid the risk of breaking something? I'm happy to adjust this to delay throwing "invalid-request" until the body is actually used.

File: services/galley/src/Galley/API/LegalHold.hs (hunk 1 above)
Comment (fisx) [reply services/galley/src/Galley/API/LegalHold.hs]: I also wonder if we can't just return `userLHStatus` below, and not introduce a new type in the first place?  That would shift exception handling around, not sure what I think of that either.

I say if in doubt, don't change this at all!  :)

File: services/galley/src/Galley/API/LegalHold.hs [hunk 2]
```diff
@@ -204,10 +222,10 @@ approveDevice (zusr ::: tid ::: uid ::: connId ::: req ::: _) = do
   legalHoldAuthToken <- Client.getLegalHoldAuthToken uid mPassword
   LHService.confirmLegalHold clientId tid uid legalHoldAuthToken
   LegalHoldData.setUserLegalHoldStatus tid uid UserLegalHoldEnabled
-  -- TODO: send event at this point (see also:
-  -- https://github.com/wireapp/wire-server/pull/802#pullrequestreview-262280386)
-  pure empty
   where
+    -- TODO: send event at this point (see also:
```
Comment (fisx) [inline services/galley/src/Galley/API/LegalHold.hs]: This comment should go one line up, to where we're returning from this function.

File: services/galley/src/Galley/API/LegalHold.hs (hunk 1 above)
Comment (mheinzel) [reply services/galley/src/Galley/API/LegalHold.hs]: Makes sense, there also some more types like this here and in brig.

Ideally, there would be some kind of `galley-api` package where we put those types and routes and also the mapping to status codes you see above. `galley-types` is kind of like that, but it doesn't seem like the status code mapping fits there are the moment.

File: services/galley/src/Galley/API/LegalHold.hs (hunk 2 above)
Comment (mheinzel) [reply services/galley/src/Galley/API/LegalHold.hs]: ah, seems like ormolu pulls this into the `where`, so I have to put it two lines up to keep it from doing that. Or I add a `pure ()` below. :D

File: services/galley/src/Galley/API/Teams.hs [hunk 3]
```diff
@@ -391,10 +407,20 @@ updateTeamMember (zusr ::: zcon ::: tid ::: req ::: _) = do
   -- push to all members (user is privileged)
   let pushPriv = newPush zusr (TeamEvent ePriv) $ privilegedRecipients
   for_ pushPriv $ \p -> push1 $ p & pushConn .~ Just zcon
-  pure empty
 
-deleteTeamMember :: UserId ::: ConnId ::: TeamId ::: UserId ::: Request ::: Maybe JSON ::: JSON -> Galley Response
-deleteTeamMember (zusr ::: zcon ::: tid ::: remove ::: req ::: _ ::: _) = do
+deleteTeamMemberH :: UserId ::: ConnId ::: TeamId ::: UserId ::: Request ::: Maybe JSON ::: JSON -> Galley Response
+deleteTeamMemberH (zusr ::: zcon ::: tid ::: remove ::: req ::: _ ::: _) = do
+  body <- fromJsonBody (JsonRequest req) -- TODO: can we demand it here?
```
[the remark this answers is not in the document]
Comment (mheinzel) [reply services/galley/src/Galley/API/Teams.hs]: Then let's do that.

File: services/galley/src/Galley/API/LegalHold.hs (hunk 1 above)
Comment (mheinzel) [reply services/galley/src/Galley/API/LegalHold.hs]: So should we leave it for now?

File: services/galley/src/Galley/API/Update.hs [hunk 4]
```diff
@@ -505,22 +529,28 @@ newMessage usr con cnv msg now (m, c, t) ~(toBots, toUsers) =
...
 
-updateConversationName :: UserId ::: ConnId ::: ConvId ::: JsonRequest ConversationRename -> Galley Response
-updateConversationName (zusr ::: zcon ::: cnv ::: req) = do
-  body <- fromJsonBody req
+updateConversationNameH :: UserId ::: ConnId ::: ConvId ::: JsonRequest ConversationRename -> Galley Response
+updateConversationNameH (zusr ::: zcon ::: cnv ::: req) = do
+  convRename <- fromJsonBody req
+  updateConversationName zusr zcon cnv convRename
+
+updateConversationName :: UserId -> ConnId -> ConvId -> ConversationRename -> Galley Response
```
Comment (fisx) [inline services/galley/src/Galley/API/Update.hs]:
```suggestion
updateConversationName :: UserId -> ConnId -> ConvId -> ConversationRename -> Galley ()
```

File: services/galley/src/Galley/API/Update.hs [hunk 5]
```diff
@@ -295,61 +305,66 @@ addCode (usr ::: zcon ::: cnv) = do
   mCode <- Data.lookupCode key ReusableCode
   case mCode of
     Nothing -> do
-      c <- generate cnv ReusableCode (Timeout 3600 * 24 * 365) -- one year TODO: configurable
-      Data.insertCode c
+      code <- generate cnv ReusableCode (Timeout 3600 * 24 * 365) -- one year TODO: configurable
```
Comment (fisx) [inline services/galley/src/Galley/API/Update.hs]: thanks!  :)

File: services/galley/src/Galley/API/Update.hs [hunk 6]
```diff
@@ -578,6 +600,8 @@ newMessage usr con cnv msg now (m, c, t) ~(toBots, toUsers) =
                   . set pushTransient (newOtrTransient msg)
            in (toBots, p : toUsers)
 
+--
```
Comment (fisx) [inline services/galley/src/Galley/API/Update.hs]: ?

File: services/galley/src/Galley/API/Update.hs [hunk 7]
```diff
@@ -279,6 +280,8 @@ updateConversationMessageTimer usr zcon cnv timerUpdate@(ConversationMessageTime
       pushEvent timerEvent users bots zcon
       pure timerEvent
 
+--
```
Comment (fisx) [inline services/galley/src/Galley/API/Update.hs]: ?

File: services/galley/src/Galley/API/Update.hs [hunk 8]
```diff
@@ -649,32 +653,38 @@ isTyping zusr zcon cnv typingData = do
         & pushRoute .~ RouteDirect
         & pushTransient .~ True
 
-addService :: JsonRequest Service -> Galley Response
-addService req = do
+addServiceH :: JsonRequest Service -> Galley Response
+addServiceH req = do
+  -- simple enough to not need a separate function
```
Comment (fisx) [inline services/galley/src/Galley/API/Update.hs]:
```suggestion
```

(we do that in other places, and we don't want to comment on it every time.)

File: services/galley/src/Galley/API/Update.hs [hunk 9]
```diff
@@ -649,32 +653,38 @@ isTyping zusr zcon cnv typingData = do
...
   Data.insertService =<< fromJsonBody req
   return empty
 
-rmService :: JsonRequest ServiceRef -> Galley Response
-rmService req = do
+rmServiceH :: JsonRequest ServiceRef -> Galley Response
+rmServiceH req = do
+  -- simple enough to not need a separate function
```
Comment (fisx) [inline services/galley/src/Galley/API/Update.hs]:
```suggestion
```

File: services/galley/src/Galley/API/CustomBackend.hs [hunk 10]
```diff
@@ -19,21 +19,27 @@ import Network.Wai.Utilities
...
   customBackend <- fromJsonBody req
+  -- simple enough to not need a separate function
   Data.setCustomBackend domain customBackend
   pure (empty & setStatus status201)
 
-internalDeleteCustomBackendByDomain :: EmailDomain ::: JSON -> Galley Response
-internalDeleteCustomBackendByDomain (domain ::: _) = do
+internalDeleteCustomBackendByDomainH :: EmailDomain ::: JSON -> Galley Response
+internalDeleteCustomBackendByDomainH (domain ::: _) = do
+  -- simple enough to not need a separate function
```
Comment (fisx) [inline services/galley/src/Galley/API/CustomBackend.hs]:
```suggestion
```

**Pull request merged**